File: backend/core-api/src/main/java/ai/govbiz/core/client/bizinfo/BizInfoClientException.java
package ai.govbiz.core.client.bizinfo;

public final class BizInfoClientException extends RuntimeException {

    public enum Failure {
        NOT_CONFIGURED,
        UPSTREAM_ERROR,
        INVALID_RESPONSE,
        UNAVAILABLE,
        TIMEOUT
    }

    private final Failure failure;

    private BizInfoClientException(Failure failure, String message, Throwable cause) {
        super(message, cause);
        this.failure = failure;
    }

    public static BizInfoClientException notConfigured() {
        return new BizInfoClientException(
                Failure.NOT_CONFIGURED,
                "BizInfo service key is not configured",
                null);
    }

    public static BizInfoClientException upstreamError(String message, Throwable cause) {
        return new BizInfoClientException(Failure.UPSTREAM_ERROR, message, cause);
    }

    public static BizInfoClientException invalidResponse(String message, Throwable cause) {
        return new BizInfoClientException(Failure.INVALID_RESPONSE, message, cause);
    }

    public static BizInfoClientException unavailable(Throwable cause) {
        return new BizInfoClientException(
                Failure.UNAVAILABLE,
                "BizInfo API could not be reached",
                cause);
    }

    public static BizInfoClientException timeout(Throwable cause) {
        return new BizInfoClientException(
                Failure.TIMEOUT,
                "BizInfo API request timed out",
                cause);
    }

    public Failure failure() {
        return failure;
    }
}

File: backend/core-api/src/main/kotlin/ai/govbiz/core/client/bizinfo/BizInfoClientException.kt
package ai.govbiz.core.client.bizinfo

class BizInfoClientException private constructor(
    val failure: Failure,
    message: String,
    cause: Throwable?,
) : RuntimeException(message, cause) {

    enum class Failure {
        NOT_CONFIGURED,
        UPSTREAM_ERROR,
        INVALID_RESPONSE,
        UNAVAILABLE,
        TIMEOUT,
    }

    fun failure(): Failure = failure

    companion object {
        @JvmStatic
        fun notConfigured(): BizInfoClientException =
            BizInfoClientException(
                Failure.NOT_CONFIGURED,
                "BizInfo service key is not configured",
                null,
            )

        @JvmStatic
        fun upstreamError(message: String, cause: Throwable?): BizInfoClientException =
            BizInfoClientException(Failure.UPSTREAM_ERROR, message, cause)

        @JvmStatic
        fun invalidResponse(message: String, cause: Throwable?): BizInfoClientException =
            BizInfoClientException(Failure.INVALID_RESPONSE, message, cause)

        @JvmStatic
        fun unavailable(cause: Throwable?): BizInfoClientException =
            BizInfoClientException(
                Failure.UNAVAILABLE,
                "BizInfo API could not be reached",
                cause,
            )

        @JvmStatic
        fun timeout(cause: Throwable?): BizInfoClientException =
            BizInfoClientException(
                Failure.TIMEOUT,
                "BizInfo API request timed out",
                cause,
            )
    }
}
